var/spack/repos/builtin/packages/scotch/package.py

```diff
@@ -37,6 +37,8 @@ class Scotch(CMakePackage, MakefilePackage):
     version("5.1.10b", sha256="54c9e7fafefd49d8b2017d179d4f11a655abe10365961583baaddc4eeb6a9add")
 
     build_system(conditional("cmake", when="@7:"), "makefile", default="cmake")
+    variant("threads", default=True, description="use POSIX Pthreads within Scotch and PT-Scotch")
+    variant("mpi_thread", default=True, description="use multi-threaded algorithms")
```

Comment:
Shouldn't this variant restrict which MPI implementations are used, since not every MPI library is configured to support `MPI_THREAD_MULTIPLE`?

Reply:
thanks for this info, @pghysels. can we just set threads and mpi_threads to `False`? users who wish to turn them on can just add `+threads` or `+mpi_threads` since they will know which mpi they are using. if you and @alalazo approve,i can make those changes.

as for the virtual mpi dependency, can you clarify what you mean? is that something you would like me to add?

Reply:
You can add mpi as a dependency, which can be provided by different packages. I'm not a spack developer, but I thought it would be nice if spack had a `thread_multiple` variant for this (virtual) mpi dependency, then this can be enabled/disabled in openmpi, and in the mpi libraries that don't support it there would be a conflict. Then every package that requires `thread_multiple` doesn't have to check all mpi implementations.

I'm ok with setting `mpi_threads` to `False`.

Reply:
ok i see what you mean. i think the issue here would be that, if we rely on the openmpi (or other MPI) package script to set `thread_multiple` rather than explicitly set it in the scotch script, the `thread_multiple` setting from openmpi would propagate throughout the stack, instead of just being applied to the scotch build. for example, someone may want `thread_multiple` set to false for scotch, but true in other library builds. perhaps i am misunderstanding, but it seems to me that it would be easier for now to just control the thread/mpi_thread settings for scotch on the scotch package.py level.

my simple proposal would just be the change to:

```
variant("threads", default=False description="use POSIX Pthreads within Scotch and PT-Scotch")
variant("mpi_thread", default=False, description="use multi-threaded algorithms"
```

 i am also not a spack developer, so if you would like to suggest an alternative, please feel free :)

Reply:
Isn't this supposed to propagate through the stack? If you want to use scotch with `+mpi_thread`, then it should propagate to the mpi library to make sure mpi is configured with support for it, or trigger a conflict.

It's ok to disable it yes.  But I think adding something like the following  is needed to make it work correctly with `+mpi_thread` when using openmpi:
```
    depends_on("openmpi +threads_multiple", when="+mpi +mpi_thread ^openmpi")
```

Reply:
Isn't this supposed to propagate through the stack? If you want to use scotch with `+mpi_thread`, then it should propagate to the mpi library to make sure mpi is configured with support for it, or trigger a conflict.

It's ok to disable it yes.  But I think adding something like the following  is needed to make it work correctly with `+mpi_thread` when using openmpi:
```
    depends_on("openmpi +threads_multiple", when="+mpi +mpi_thread ^openmpi")
```

Reply:
ah yes, that makes sense. i was just imagining a situation in which a scotch specific cmake flag would be applied to all packages that required openmpi.

i changed the `mpi_thread` default to `False` and added the `depends_on` statement you recommended.

Reply:
Now I see that the `openmpi +threads_multiple` variant is only available for `openmpi@1.5.4:2` and the current version is `4.1.5`. So perhaps this extra `depends_on` isn't super useful.

Reply:
we can remove it then. i built scotch w/ openmpi@4.1.5 (without the depends_on) and it built ok w/ the scotch `mpi_thread` variant.